src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -139,26 +156,39 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
             }
             else
             {
-                // We're in an end tag helper block.
+                // We're in an end tag block.
 
-                var tagNameScope = _tagStack.Count > 0 ? _tagStack.Peek().TagName : string.Empty;
-
-                // Validate that our end tag helper matches the currently scoped tag helper, if not we
-                // need to error.
+                // Validate that our end tag matches the currently scoped tag, if not we may need to error.
                 if (tagNameScope.Equals(tagName, StringComparison.OrdinalIgnoreCase))
                 {
-                    ValidTagStructure(tagName, tagBlock, context);
+                    // If there are additional end tags required before we can build our block it means we're in a
+                    // situation like this: <myth req="..."><myth></myth></myth> where we're at the inside </myth>.
+                    if (currentTagBlockTracker.OpenMatchingTags > 0)
+                    {
+                        currentTagBlockTracker.OpenMatchingTags--;
+
+                        return false;
+                    }
+
+                    ValidateTagStructure(tagName, tagBlock, context);
 
                     BuildCurrentlyTrackedTagHelperBlock();
                 }
                 else
                 {
+                    // If there are not TagHelperDescriptors associated with the end tag block that also have no 
+                    // required attributes then it means we can't be a TagHelper, bail out.
+                    if (!_provider.GetDescriptors(tagName, attributeNames: Enumerable.Empty<string>()).Any())
+                    {
+                        return false;
+                    }
```

Comment:
on second thought, these lines break recovery.  could be in a situation like
``` html
<tagWithRequiredAttribute attribute="value">
  <anotherTagHelper>
  </tagWithRequiredAttribute>
```
when parsing the end tag, will look for a tag helper targeting `tagWithRequiredAttribute` with no attributes.  can't recover because the helper actually requires one (provided) attribute.

should at least let user know the `<anotherTagHelper>` element is missing its end tag or that Razor can't match the `</tagWithRequiredAttribute>` end tag with its start tag.